keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1273 @@
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+- 2023-01-23: Initial version of the KEP PR [Backoff Limit Per Job #3774](https://github.com/kubernetes/enhancements/pull/3774)
+- 2023-04-26: The KEP PR [Backoff limit per Job Index #3967](https://github.com/kubernetes/enhancements/pull/3967) takes over from [#3774](https://github.com/kubernetes/enhancements/pull/3774)
+- 2023-05-08: The KEP PR ready for review
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Alternative names
+
+Alternative names proposed for the structure to wrap the `backoffLimitPerIndex`
+and the `maxFailedIndexes` fields:
+1. `runPolicy`
+2. `completionPolicy`
+3. `failurePolicy`
+
+The option (3.) suggests that the fields are about declaring the Job as failed.
+However, the `backoffLimitPerIndex` field not only allows to count failures
+towards the backoff limit per index, but also allows all indexes to execute
+despite failures, thus more generic names, like (1.) and (2.) are preferred.
+
+Also the options (1.) and (2.) may be reused in the context of success policy
+which is subject of
+[Job success/completion policy](https://github.com/kubernetes/kubernetes/issues/117600).
+It might be beneficial for the API to consider the conditions for the Job
+success or failure under the same field.
+
+### Mark Job Complete if some indexes failed
+
+The alternative to the proposed [Job completion](#job-completion) strategy.
+
+Allow execution of all indexes, up to `.spec.runPolicy.maxFailedIndexes` of
+failed indexes. Then, mark the Job `Complete` even if some indexes failed.
+The Job is marked `Failed` only if the number of failed indexes exceeds the
+specified `.spec.runPolicy.maxFailedIndexes` limit, in that case, the `reason`
+field could be `FailedIndexes`, and the `message` field would list the failed
+indexes up to a couple of them.
+
+**Reasons for deferring / rejecting**
+
+This approach is less intuitive to the end-users of the API, compared
+to the proposal. In particular, in some cases it would require custom logic in
+the user's controller to determine if the Job is failed.
+
+### Support backoffLimitPerIndex when restartPolicy=OnFailure
+
+We've considered supporting the backoffLimitPerIndex when pod's `restartPolicy=OnFailure`.
+However, in this mode it is Kubelet's responsibility to restart the pod, while
+the maximal number of restarts would be enforced by the Job controller, which
+is racy in its nature. For example, in-between the checks by the Job controller,
+Kubelet execute more restarts than the specified `.spec.backoffLimit`.
```

Comment:
I don't think kubelet will execute more restarts than expected, but they do have race condition here.

Reply:
Let say Job's `backoffLimit=0`, and `restartPolicy: OnFailure`. Then, kubelet would restart the pod containers because it is unaware that Job's backoffLimit=0. Similar unncessary restarts can happen for higher backoffLimit.

It is then Job controller that enforces the limit on the job sync. We could use the similar strategy for backoffLimitPerIndex, but it seems better to delegate controlling the restarts to kubelet, by maxRestartTimes.